dev/src/dev/rewrite_defendpoint_1.clj

```diff
@@ -0,0 +1,422 @@
+  (or (when-let [as (z/find (z/down (z/of-node fn-args))
+                            #(= (z/sexpr %) :as))]
+        (let [request-map (z/right as)]
+          (assert (= #{:body}
+                     (set (vals (z/sexpr request-map))))
+                  (format "Don't know how to rewrite request map with non-:body keys: %s" (pr-str request-map)))
+          (when-let [body-key (z/find (z/down request-map)
+                                      #(= (z/sexpr %) :body))]
+            (z/node (z/left body-key)))))
+      (n/token-node '_body)))
+
+(mu/defn- body-schema :- [:maybe ::node]
+  [{:keys [schema-map], :as parsed} :- ::parsed]
+  (when schema-map
+    (if (= (n/tag (body-binding parsed)) :token)
+      ;; body is a plain symbol, do not build a the usual map schema.
+      (get schema-map (n/sexpr (body-binding parsed)))
+      ;; otherwise body is a destructured map
+      (let [body-keys (->> schema-map
+                           keys
+                           (remove (set (route-args-symbols parsed)))
+                           (remove (set (query-args-symbols parsed))))]
+        (schema-map->malli (select-keys* schema-map body-keys))))))
+
+(mu/defn- new-body-arg-nodes :- [:maybe [:sequential ::node]]
+  [parsed :- ::parsed]
+  (when-not (unused-binding? (body-binding parsed))
+    (list*
+     (n/newline-node "\n")
+     (n/whitespace-node "   ")
+     (body-binding parsed)
+     (schema-specifier (body-schema parsed)))))
+
+;;;
+;;; rewriting defendpoint symbol
+;;;
+
+(mu/defn- rewrite-defendpoint-symbol :- ::zloc
+  [_parsed     :- ::parsed
+   defendpoint :- ::zloc]
+  (z/replace (z/down defendpoint) (n/token-node 'api.macros/defendpoint)))
+
+;;;
+;;; rewrite method
+;;;
+
+(mu/defn- rewrite-method :- ::zloc
+  [_parsed defendpoint :- ::zloc]
+  (let [method (-> defendpoint z/down z/right)]
+    (z/replace method (n/keyword-node (keyword (u/lower-case-en (name (z/sexpr method))))))))
+
+;;;
+;;; rewrite argslist
+;;;
+
+(mu/defn- find-argslist :- [:and
+                            ::zloc
+                            [:fn
+                             {:error/message "zipper location pointing to vector node"}
+                             #(= (z/tag %) :vector)]]
+  [defendpoint :- ::zloc]
+  (let [route (-> defendpoint
+                  z/down    ; symb
+                  z/right   ; method
+                  z/right)] ; route
+    ;; first vector after route
+    (z/find-next route #(= (z/tag %) :vector))))
+
+(mu/defn- new-args-node :- ::node
+  [parsed :- ::parsed]
+  (n/vector-node
+   (into []
+         (mapcat (fn [f]
+                   (f parsed)))
+         [new-route-param-arg-nodes
+          new-query-param-arg-nodes
+          new-body-arg-nodes])))
+
+(mu/defn- rewrite-args :- ::zloc
+  [parsed      :- ::parsed
+   defendpoint :- ::zloc]
+  (z/replace (find-argslist defendpoint) (new-args-node parsed)))
+
+;;;
+;;; Add metadata map if needed
+;;;
+
+(mu/defn- add-metadata-map :- ::zloc
```

Comment:
Old `defendpoint` could (rarely) have stuff like `^:deprecated GET "/:id"` or `^:multipart POST ...`, since the new version uses keywords (which can't have metadata on them) we now can specify a metadata map after the docstring, before the argslist, like `defn` or `defmulti`.